internal/pubsub/pubsub.go

```diff
@@ -61,6 +61,24 @@ func (m *Message) Ack() {
 	m.isAcked = true
 }
 
+// Value returns a MessageValue based on the Message receiver.
+func (m *Message) Value() MessageValue {
+	return MessageValue{
+		Body:     string(m.Body),
+		Metadata: m.Metadata,
+	}
+}
+
+// MessageValue contains the exported fields of a Message with a UTF8 Body, making it
```

Comment:
This just feels like overkill to me -- a whole new exported struct type just so diffs look nicer in our tests? It's not obvious to me that it will be useful to end users at all; they may want to log Messages, but it's not obvious that MessageValue.String() will produce something useful.

Why not a best-effort `String()` function that does something like:
- Take the first N bytes of `Body`
  - If it's a string, print it, with "... (%d bytes total)" appended if len(Body) > N.
  - If it's not a string, print "body with binary data (%d bytes)".
- Append the number of metadata keys (e.g., "(%d metadata keys)") if len(Metadata) > 0, or maybe a string representation of the metadata if it's short enough.

The test could capture a slice of both `*Message` and strings from `Message.String()`, and compare the strings first (resulting in a nice diff if they are not the same), and the actual messages (hiding unexported fields) second (resulting in the existing diff if `String()` returned the same thing but the underlying `Message`s are different; this should be extremely unlikely for the short messages used in the test).

Reply:
Yes, I think the improvement in readability of test output is well worth the cost of an additional exported type.

Why is it not obvious that `MessageValue.String()` will produce something useful? Its fields are the ones with the data users have provided and presumably care about.

The main reason not add a best-effort `String()` method on Message is because it fails to change the output of `cmp.Diff`, which was what motivated this change in the first place. I don't think it would give a very readable result to use `cmp.Diff` on a string that contains a bunch of message `String`s.

Also `String` is not very flexible, forcing you to choose between breaking "%v" or "%+v". It's possible to implement `Formatter` but the code for that is more complex and harder to test. I got partway through doing that and wasn't sure if I had covered all the cases. With this PR, all the cases are covered.

Reply:
@ijt I strongly disagree. We should not be changing the API just for tests. I might understand if it was necessary to exercise some behavior that we couldn't otherwise, but this is just cosmetic.

Reply:
Why not? Testability is important, and having readable output when tests fail is also important.

Reply:
I think the difference is the size of the audience. The number of end users exposed to the exported API will (hopefully) be orders of magnitude larger than the users exposed to the test code.

Personally I don't find readable output when tests fail super high on my list of priorities. When the test fails, you end up in the code debugging anyway, and if it's not obvious what the problem is you can usually easily add code to make it obvious. I'd rather have the test be simple and straightforward.

I still find it interesting to consider how to make it easy/useful for users to log/print a `Message`, but for me at least the impact on the test is secondary.

Reply:
Ah, if it's a low priority then I won't spend any more time on it. Closing this.